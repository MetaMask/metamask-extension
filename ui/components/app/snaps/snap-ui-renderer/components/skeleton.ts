import { SkeletonElement } from '@metamask/snaps-sdk/jsx';
import { BorderRadius } from '../../../../../helpers/constants/design-system';
import { mapSnapBorderRadiusToExtensionBorderRadius } from '../utils';
import { UIComponentFactory } from './types';

const DEFAULT_SKELETON_WIDTH = '100%';
const DEFAULT_SKELETON_HEIGHT = 22;
const DEFAULT_SKELETON_BORDER_RADIUS = BorderRadius.MD;

export const skeleton: UIComponentFactory<SkeletonElement> = ({ element }) => {
  return {
    element: 'Skeleton',
    props: {
      width: element.props.width ?? DEFAULT_SKELETON_WIDTH,
      height: element.props.height ?? DEFAULT_SKELETON_HEIGHT,
      borderRadius: element.props.borderRadius
        ? mapSnapBorderRadiusToExtensionBorderRadius(element.props.borderRadius)
        : DEFAULT_SKELETON_BORDER_RADIUS,
    },
  };
};
