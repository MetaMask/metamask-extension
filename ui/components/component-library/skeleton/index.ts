export { Skeleton } from './skeleton';
export type {
  SkeletonProps,
  SkeletonComponent,
  SkeletonStyleUtilityProps,
} from './skeleton.types';
