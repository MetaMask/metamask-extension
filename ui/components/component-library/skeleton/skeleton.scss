.mm-skeleton {
  @keyframes pulse {
    50% {
      opacity: 0.5;
    }
  }

  animation: pulse 2s cubic-bezier(0.4, 0, 0.6, 1) infinite;
  background-clip: padding-box;
  cursor: default;
  pointer-events: none;
  user-select: none;

  &--hide-children * {
    visibility: hidden;
  }
}
