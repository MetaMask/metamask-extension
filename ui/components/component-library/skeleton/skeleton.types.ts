import React from 'react';

import type {
  StyleUtilityProps,
  PolymorphicComponentPropWithRef,
} from '../box';

export interface SkeletonStyleUtilityProps
  extends Omit<StyleUtilityProps, 'height' | 'width'> {
  /**
   * Additional className to add to the Skeleton
   */
  className?: string;
  /**
   * The height of the Skeleton
   */
  height?: number | string;
  /**
   * The width of the Skeleton
   */
  width?: number | string;
  /**
   * The children of the Skeleton
   */
  children?: React.ReactNode;
}

export type SkeletonProps<C extends React.ElementType> =
  PolymorphicComponentPropWithRef<C, SkeletonStyleUtilityProps>;

export type SkeletonComponent = <C extends React.ElementType = 'span'>(
  props: SkeletonProps<C>,
) => React.ReactElement | null;
