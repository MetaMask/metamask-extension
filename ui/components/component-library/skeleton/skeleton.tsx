import React from 'react';
import classnames from 'classnames';

import {
  BackgroundColor,
  BorderRadius,
} from '../../../helpers/constants/design-system';
import { Box } from '..';
import type { PolymorphicRef, BoxProps } from '../box';

import { SkeletonProps, SkeletonComponent } from './skeleton.types';

export const Skeleton: SkeletonComponent = React.forwardRef(
  <C extends React.ElementType = 'div'>(
    {
      className = '',
      height,
      width,
      children,
      hideChildren,
      ...props
    }: SkeletonProps<C>,
    ref?: PolymorphicRef<C>,
  ) => (
    <Box
      className={classnames(
        'mm-skeleton',
        {
          'mm-skeleton--hide-children': hideChildren,
        },
        className,
      )}
      backgroundColor={BackgroundColor.iconMuted}
      borderRadius={BorderRadius.SM}
      ref={ref}
      {...(props as BoxProps<C>)}
      style={{ ...props?.style, height, width }}
    >
      {children}
    </Box>
  ),
);

export default Skeleton;
