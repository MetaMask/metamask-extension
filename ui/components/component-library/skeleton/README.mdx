import { Controls, Canvas } from '@storybook/blocks';

import * as SkeletonStories from './skeleton.stories';

# Skeleton

`Skeleton` is a placeholder component that is used while the content is loading.

<Canvas of={SkeletonStories.DefaultStory} />

## Props

<Controls of={SkeletonStories.DefaultStory} />

### Width Height

Use the `width` and `height` props to specify the width and height of the `Skeleton` component. It accepts both numbers and strings.

<Canvas of={SkeletonStories.WidthHeight} />

```jsx
import { Skeleton } from '../../component-library';

<Skeleton height={32} width={300} />
<Skeleton height={16} width={250} />
<Skeleton height={16} width={250} />
```

### Children

Use the `children` prop to pass the `Skeleton` children.

This is useful to create a `Skeleton` component that has a similar layout to the content that will be loaded.

<Canvas of={SkeletonStories.Children} />

```jsx
import {
  Display,
  FlexDirection,
  BackgroundColor,
  BorderRadius,
} from '../../../helpers/constants/design-system';
import { Skeleton } from '../../component-library';

<Skeleton
  display={Display.Flex}
  flexDirection={FlexDirection.Column}
  gap={4}
  backgroundColor={BackgroundColor.backgroundAlternative}
  borderRadius={BorderRadius.LG}
  padding={4}
>
  <Skeleton height={32} width="100%" />
  <Skeleton height={16} width="95%" />
  <Skeleton height={16} width="95%" />
</Skeleton>;
```

### Hide Children

Use the `hideChildren` prop to hide the `children` of the `Skeleton` component. This is useful to make sure the `Skeleton` component is the same size as the content that will be loaded.

<Canvas of={SkeletonStories.HideChildren} />

```jsx
import { Skeleton, Text, TextVariant } from '../../component-library';

isLoaded ? (
  <Text variant={TextVariant.headingMd}>Content to load</Text>
) : (
  <Skeleton width="max-content" hideChildren={true}>
    <Text variant={TextVariant.headingMd}>Hidden placeholder text</Text>
  </Skeleton>
);
```

### Border Radius

Use the `borderRadius` prop and `BorderRadius` enum to specify the border radius of the `Skeleton` component.

<Canvas of={SkeletonStories.BorderRadiusStory} />

```jsx
import { BorderRadius } from '../../../helpers/constants/design-system'
import { Skeleton } from '../../component-library';

<Skeleton borderRadius={BorderRadius.full} height={32} width={32} />
<Skeleton borderRadius={BorderRadius.pill} height={32} width="100%" />
<Skeleton borderRadius={BorderRadius.LG} height={32} width="100%" />
<Skeleton borderRadius={BorderRadius.MD} height={32} width="100%" />
<Skeleton borderRadius={BorderRadius.SM} height={32} width="100%" />
<Skeleton borderRadius={BorderRadius.XS} height={32} width="100%" />
```
