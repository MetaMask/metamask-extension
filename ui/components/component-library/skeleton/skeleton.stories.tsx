import React from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import {
  BackgroundColor,
  BlockSize,
  Display,
  FlexDirection,
  BorderRadius,
  TextVariant,
  AlignItems,
} from '../../../helpers/constants/design-system';
import { Box, Button, ButtonVariant, Text } from '..';
import README from './README.mdx';
import { Skeleton } from './skeleton';

const meta: Meta<typeof Skeleton> = {
  title: 'Components/ComponentLibrary/Skeleton',
  component: Skeleton,
  parameters: {
    docs: {
      page: README,
    },
  },
  argTypes: {
    className: {
      control: 'text',
    },
  },
  args: {
    className: '',
  },
};

export default meta;
type Story = StoryObj<typeof Skeleton>;

export const DefaultStory: Story = {
  args: {
    height: 32,
  },
};

DefaultStory.storyName = 'Default';

export const WidthHeight: Story = {
  render: () => (
    <Box display={Display.Flex} flexDirection={FlexDirection.Column} gap={4}>
      <Skeleton height={32} width={300} />
      <Skeleton height={16} width={250} />
      <Skeleton height={16} width={250} />
    </Box>
  ),
};

export const Children: Story = {
  render: () => (
    <Skeleton
      display={Display.Flex}
      flexDirection={FlexDirection.Column}
      gap={4}
      backgroundColor={BackgroundColor.backgroundAlternative}
      borderRadius={BorderRadius.LG}
      padding={4}
    >
      <Skeleton height={32} width="100%" />
      <Skeleton height={16} width="95%" />
      <Skeleton height={16} width="95%" />
    </Skeleton>
  ),
};

export const HideChildren: Story = {
  args: {
    width: 'max-content',
    hideChildren: true,
  },
  render: (args) => {
    const [isLoaded, setIsLoaded] = React.useState(false);
    return (
      <Box>
        <Button
          onClick={() => setIsLoaded(!isLoaded)}
          variant={ButtonVariant.Secondary}
          marginBottom={4}
        >
          Toggle isLoading
        </Button>
        {isLoaded ? (
          <div>
            <Text variant={TextVariant.headingMd}>Content to load</Text>
          </div>
        ) : (
          <Skeleton {...args}>
            <Text variant={TextVariant.headingMd}>Hidden placeholder text</Text>
          </Skeleton>
        )}
      </Box>
    );
  },
};

export const BorderRadiusStory: Story = {
  render: () => (
    <Box display={Display.Flex} flexDirection={FlexDirection.Column} gap={4}>
      <Box display={Display.Flex} alignItems={AlignItems.center} gap={4}>
        <Text>BorderRadius.full</Text>
        <Skeleton borderRadius={BorderRadius.full} height={32} width={32} />
      </Box>
      <Box display={Display.Flex} alignItems={AlignItems.center} gap={4}>
        <Text>BorderRadius.full</Text>
        <Skeleton borderRadius={BorderRadius.pill} height={32} width="100%" />
      </Box>
      <Box display={Display.Flex} alignItems={AlignItems.center} gap={4}>
        <Text>BorderRadius.LG</Text>
        <Skeleton borderRadius={BorderRadius.LG} height={32} width="100%" />
      </Box>
      <Box display={Display.Flex} alignItems={AlignItems.center} gap={4}>
        <Text>BorderRadius.MD</Text>
        <Skeleton borderRadius={BorderRadius.MD} height={32} width="100%" />
      </Box>
      <Box display={Display.Flex} alignItems={AlignItems.center} gap={4}>
        <Text>BorderRadius.SM</Text>
        <Skeleton borderRadius={BorderRadius.SM} height={32} width="100%" />
      </Box>
      <Box display={Display.Flex} alignItems={AlignItems.center} gap={4}>
        <Text>BorderRadius.XS</Text>
        <Skeleton borderRadius={BorderRadius.XS} height={32} width="100%" />
      </Box>
    </Box>
  ),
};

BorderRadiusStory.storyName = 'BorderRadius';
