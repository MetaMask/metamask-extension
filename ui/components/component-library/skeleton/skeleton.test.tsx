import React from 'react';
import { render } from '@testing-library/react';

import { Skeleton } from './skeleton';

describe('Skeleton', () => {
  it('should render Skeleton without error', () => {
    const { getByTestId } = render(<Skeleton data-testid="skeleton" />);
    expect(getByTestId('skeleton')).toBeDefined();
    expect(getByTestId('skeleton')).toHaveClass('mm-skeleton');
  });
  it('should match snapshot', () => {
    const { container } = render(<Skeleton />);
    expect(container).toMatchSnapshot();
  });
  it('should render with and additional className', () => {
    const { getByTestId } = render(
      <Skeleton data-testid="skeleton" className="test-class" />,
    );
    expect(getByTestId('skeleton')).toHaveClass('test-class');
  });
});
