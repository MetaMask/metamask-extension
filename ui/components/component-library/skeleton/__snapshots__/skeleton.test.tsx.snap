// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton should match snapshot 1`] = `
<div>
  <div
    class="mm-box mm-skeleton mm-box--padding-4 mm-box--background-color-icon-muted mm-box--rounded-sm"
  />
</div>
`;
